torch/quantization/_equalize.py

```diff
@@ -0,0 +1,133 @@
+    temp[axis] = size_of_tensor_dim - 1
+
+    mins = mins.permute(temp)
+    maxs = maxs.permute(temp)
+
+    # minimizing over all axises except the last axis
+    for i in range(size_of_tensor_dim - 1):
+        mins = torch.min(mins, 0)[0]
+        maxs = torch.max(maxs, 0)[0]
+
+    assert(mins.size()[0] == input.size()[axis])
+    return maxs - mins
+
+def scaling_channels(module1, module2, output_axis=0, input_axis=1):
+    ''' Given two adjacent tensors', the weights are scaled such that
+    the ranges of the first tensors' output channel are equal to the
+    ranges of the second tensors' input channel
+
+    '''
+    tensor1 = module1.weight
+    tensor2 = module2.weight
+    bias = module1.bias
+
+    output_channel_tensor1 = channel_range(tensor1, output_axis)
+    input_channel_tensor2 = channel_range(tensor2, input_axis)
+
+    # producing scaling factors to applied
+    input_channel_tensor2 += 1e-9
+    scaling_factors = torch.sqrt(output_channel_tensor1 / input_channel_tensor2)
+    r_scaling_factors = torch.reciprocal(scaling_factors)
+
+    bias = bias * r_scaling_factors
+
+    # formatting the scaling (1D) tensors to be applied on the given argument tensors
+    # pads axis to (1D) tensors to then be broadcasted
+    size1 = [1 for x in range(len(tensor1.size()))]
+    size1[output_axis] = tensor1.size()[output_axis]
+    size2 = [1 for x in range(len(tensor2.size()))]
+    size2[input_axis] = tensor2.size()[input_axis]
+
+    scaling_factors = torch.reshape(scaling_factors, size2)
+    r_scaling_factors = torch.reshape(r_scaling_factors, size1)
+
+    tensor1 = tensor1 * r_scaling_factors
+    tensor2 = tensor2 * scaling_factors
+
+    module1.weight.data = tensor1
+    module1.bias.data = bias
+    module2.weight.data = tensor2
+
+def cross_layer_equalization(module1, module2):
+    ''' Given two adjacent modules, the weights are scaled such that
+    the ranges of the first modules' output channel are equal to the
+    ranges of the second modules' input channel
+
+    Scaling work is done in scaling_channels()
+    '''
+    module1_output_channel_axis = -1
+    module2_input_channel_axis = -1
+
+    # additional modules can be added, given the axises of the input/output channels are known
+    if isinstance(module1, nn.Linear) or isinstance(module1, nn.Conv2d):
+        module1_output_channel_axis = 0
+    else:
+        raise TypeError("Only Linear and Conv2d modules are supported at this time")
+    if isinstance(module2, nn.Linear) or isinstance(module2, nn.Conv2d):
+        module2_input_channel_axis = 1
+    else:
+        raise TypeError("Only Linear and Conv2d modules are supported at this time")
+
+    if module1.weight.size()[module1_output_channel_axis] != module2.weight.size()[module2_input_channel_axis]:
+        raise TypeError("Incompatible tensors")
+
+    scaling_channels(module1, module2, module1_output_channel_axis, module2_input_channel_axis)
+
+def equalize(model, paired_modules_list, threshold):
```

Comment:
would probably make sense to give a default value for `threshold`

Reply:
also, you can provide a inplace option, when inplace is false you can copy the model, and all the following modification can be inplace. example: https://github.com/pytorch/pytorch/blob/master/torch/quantization/quantize.py#L209